### PR TITLE
Option to have http.server and http.client use UNIX sockets instead of TCP

### PR DESCRIPTION
**Pull request by ChickenNuggers**
Replication and improvement of #8.

### CODE REVIEW

File: http/client.lua [hunk 1]
```diff
@@ -41,7 +41,7 @@ local function negotiate(s, options, timeout)
 				tls = default_h2_ctx or default_h1_ctx
 			end
 		end
-		local ok, err, errno = s:starttls(tls, deadline and (deadline-monotime()))
+		local ok, err, errno = s:starttls(tls, timeout)
```
Comment (daurnimator) [inline http/client.lua]: why is this line different?

File: http/client.lua [hunk 2]
```diff
@@ -91,6 +92,5 @@ local function connect(options, timeout)
 end
 
 return {
-	negotiate = negotiate;
```
Comment (daurnimator) [inline http/client.lua]: why was this line removed?

File: spec/server_spec.lua [hunk 3]
```diff
@@ -13,14 +13,24 @@ describe("http.server module", function()
...
-		}
+		local options = {}
+		if path then
+			options.path = path
+		else
+			options.host = "localhost"
+			options.port = 0
+		end
+		options.version = version
+		options.tls = tls
```
Comment (daurnimator) [inline spec/server_spec.lua]: options doesn't need/want TLS.

File: spec/server_spec.lua [hunk 4]
```diff
@@ -50,16 +64,44 @@ describe("http.server module", function()
...
-	it("works with plain http 2.0", function()
+	it("works with plain http 2.0 using IP", function()
 		simple_test(false, 2.0)
 	end);
-	(require "http.tls".has_alpn and it or pending)("works with https 2.0", function()
+	(require "http.tls".has_alpn and it or pending)("works with https 2.0 using IP", function()
 		simple_test(true, 2.0)
 	end)
+	it("works with plain http 1.1 using UNIX socket domain", function()
+		local path = os.tmpname() .. ".socket"
```
Comment (daurnimator) [inline spec/server_spec.lua]: `os.tmpname` returns a different string each time.... you're generating a random file name to delete here.

File: spec/server_spec.lua [hunk 5]
```diff
@@ -50,16 +64,44 @@ describe("http.server module", function()
...
+		simple_test(false, 1.1, os.tmpname() .. ".socket")
+		finally(function()
+			os.remove(path)
+		end)
+	end)
+	pending("works with https 1.1 using UNIX socket domain", function()
```
Comment (daurnimator) [inline spec/server_spec.lua]: why is this pending?

File: http/client.lua [hunk 6]
```diff
@@ -41,7 +41,7 @@ local function negotiate(s, options, timeout)
 				tls = default_h2_ctx or default_h1_ctx
 			end
 		end
-		local ok, err, errno = s:starttls(tls, timeout)
+		local ok, err, errno = s:starttls(tls, deadline and (deadline-monotime()))
```
Comment (daurnimator) [inline http/client.lua]: please squash this commit with the first one.

File: spec/server_spec.lua [hunk 7]
```diff
@@ -50,16 +63,58 @@ describe("http.server module", function()
...
 	end)
+	--[[
+	--
+	-- Until there is a way to generate OpenSSL contexts in this file for
+	-- UNIX domain sockets, there is no way to use TLS with this. Because
+	-- of this, the status for using TLS with UNIX domain sockets is
+	-- pending.
+	--
+	--]]
+	local socket_path
```
Comment (daurnimator) [inline spec/server_spec.lua]: What is this block doing?
You just generate a random file name... then delete it.

File: spec/server_spec.lua [hunk 8]
```diff
@@ -50,16 +63,58 @@ describe("http.server module", function()
...
+		end)
+	end)
+	it("works with plain http 2.0 using UNIX socket domain", function()
+		local path = socket_path
+		simple_test(false, 2.0, socket_path)
+		finally(function()
+			os.remove(socket_path)
+		end)
+	end); -- change first pending to 'it' when added ctx generation
+	(require "http.tls".has_alpn and pending or pending)("works with https 2.0 using UNIX socket domain", function()
```
Comment (daurnimator) [inline spec/server_spec.lua]: Get rid of the ternary, change it to just `pending`

File: spec/server_spec.lua [hunk 9]
```diff
@@ -50,16 +63,58 @@ describe("http.server module", function()
...
+			os.remove(socket_path)
+		end)
+	end)
+	it("works with plain http 2.0 using UNIX socket domain", function()
+		local path = socket_path
+		simple_test(false, 2.0, socket_path)
+		finally(function()
+			os.remove(socket_path)
+		end)
+	end); -- change first pending to 'it' when added ctx generation
```
Comment (daurnimator) [inline spec/server_spec.lua]: Why is this comment part of the above test?

File: spec/server_spec.lua (hunk 7 above)
Comment (ChickenNuggers) [inline spec/server_spec.lua]: I'm double checking now but Linux systems create the return value of os.tmpname() so I need to remove that, then use the filename with `.socket`. I'm checking later when I get home whether I actually use the path generated in the tests. I might have to create a wrapper function for this and then remove this block.

Comment (daurnimator) [inline spec/server_spec.lua]: Why bother with adding .socket? just use the result of `os.tmpname`.

File: spec/server_spec.lua (hunk 3 above)
Comment (ChickenNuggers) [inline spec/server_spec.lua]: Confirmed tonight that it does; instead it assumes (at least on my system) that it needs to generate a certificate (which breaks atm because certificates are pending).

File: spec/server_spec.lua [hunk 10]
```diff
@@ -84,4 +107,38 @@ describe("http.server module", function()
...
 	end)
+	--[[
+	--
+	-- Until there is a way to generate OpenSSL contexts in this file for
+	-- UNIX domain sockets, there is no way to use TLS with this. Because
+	-- of this, the status for using TLS with UNIX domain sockets is
+	-- pending.
+	--
+	--]]
+	local socket_path = os.tmpname()
```
Comment (daurnimator) [inline spec/server_spec.lua]: Why not generate one per test?

File: spec/server_spec.lua [hunk 11]
```diff
@@ -84,4 +107,38 @@ describe("http.server module", function()
 		assert.truthy(cq:empty())
 		assert.spy(on_stream).was.called()
 	end)
+	--[[
```
Comment (daurnimator) [inline spec/server_spec.lua]: Can you move this comment down to above the `pending` test?

File: spec/server_spec.lua [hunk 12]
```diff
@@ -84,4 +107,38 @@ describe("http.server module", function()
...
+	-- UNIX domain sockets, there is no way to use TLS with this. Because
+	-- of this, the status for using TLS with UNIX domain sockets is
+	-- pending.
+	--
+	--]]
+	local socket_path = os.tmpname()
+	os.remove(socket_path) -- in case it was generated automatically
+	it("works with plain http 1.1 using UNIX socket domain", function()
+		simple_test(false, 1.1, socket_path)
+		finally(function()
```
Comment (daurnimator) [inline spec/server_spec.lua]: you need to call `finally` before the test throws.

File: spec/server_spec.lua (hunk 10 above)
Comment (ChickenNuggers) [inline spec/server_spec.lua]: Is there a reason to? If you remove it in the finally() call, it should be done after the it() call, right?

File: spec/server_spec.lua (hunk 11 above)
Comment (ChickenNuggers) [inline spec/server_spec.lua]: Added to my todo list

File: spec/server_spec.lua (hunk 10 above)
Comment (daurnimator) [inline spec/server_spec.lua]: Yeah that's fine I guess.

File: spec/server_spec.lua [hunk 13]
```diff
@@ -4,14 +4,33 @@ describe("http.server module", function()
 	local new_headers = require "http.headers".new
 	local cqueues = require "cqueues"
 	local cs = require "cqueues.socket"
-	local function simple_test(tls, version)
+	local function assert_loop(cq, timeout)
```
Comment (daurnimator) [inline spec/server_spec.lua]: You might need to rebase again, as `assert_loop` was factored out.
